File: data_structures/1352-product-of-the-last-k-numbers.cpp
#include <bits/stdc++.h>
using namespace std;

class ProductOfNumbers {
public:
    ProductOfNumbers() {}

    void add(int num) {
        if (num == 0) {
            nums = {1};
        } else {
            nums.emplace_back(nums.back() * num);
        }
    }

    int getProduct(int k) {
        if (k >= nums.size()) { // nums.size() always > num if there is no 0s.
            return 0;
        } else {
            return nums.back() / nums[nums.size() - k - 1];
        }
    }

private:
    vector<int> nums{1};
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */

File: data_structures/1352-product-of-the-last-k-numbers.java
package data_structures;

import java.util.ArrayList;
import java.util.List;

class ProductOfNumbers {
    private List<Integer> list;

    public ProductOfNumbers() {
        this.list = new ArrayList<>();
        this.list.add(1);
    }
    
    public void add(int num) {
        if (num == 0) {
            list = new ArrayList<>();
            list.add(1);
        } else {
            int tmp = list.get(list.size() - 1) * num;
            list.add(tmp);
        }
    }
    
    public int getProduct(int k) {
        int n  = list.size();
        if (k < n)
            return list.get(n - 1) / list.get(n - k - 1);
        else 
            return 0;
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.add(num);
 * int param_2 = obj.getProduct(k);
 */